File: app/src/main/java/com/example/fristtrial/StopWatch_fragment.java
package com.example.fristtrial;

import static android.text.format.DateUtils.formatElapsedTime;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.SystemClock;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Chronometer;
import android.widget.ListView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;



import java.util.ArrayList;

public class StopWatch_fragment extends Fragment {
    public Chronometer chrono;
    public Button start, lap, clearLap;
    public ListView lapList;
    public boolean running, tstart;
    public long pauseOffset;
    public int pause1, index = 0;
    public float lap_mili;
    public ArrayAdapter<String> lapAdapter;

//    public StopwatchFragment() {
//        // Required empty public constructor
//    }

    @SuppressLint("MissingInflatedId")
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_stop_watch_fragment, container, false);
        chrono = view.findViewById(R.id.chrTimerFrag);

        lapList = view.findViewById(R.id.lapTimeFrag);
        ArrayList<String> lapArrayList = new ArrayList<>();
        lapAdapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_list_item_1, lapArrayList);
        lapList.setAdapter(lapAdapter);

        String startBtnText, stopBtnText, lapBtnText, resetBtnText;
        Context context = getActivity();
        startBtnText = context.getString(R.string.start);
        stopBtnText = context.getString(R.string.stop);
        lapBtnText = context.getString(R.string.lap);
        resetBtnText = context.getString(R.string.reset);

        start = view.findViewById(R.id.startBtnFrag);
        start.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (start.getText().equals(startBtnText)) {
                    start.setText(stopBtnText);
                    chrono.setBase(SystemClock.elapsedRealtime() - pauseOffset);
                    chrono.start();
                    running = true;
                    tstart = true;
                    lap.setText(lapBtnText);
                } else {
                    start.setText(startBtnText);
                    chrono.stop();
                    pauseOffset = SystemClock.elapsedRealtime() - chrono.getBase();
                    running = false;
                    lap.setText(resetBtnText);
                }
            }
        });


        lap = view.findViewById(R.id.lapBtnFrag);
        lap.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (lap.getText().equals(lapBtnText)) {
                    index++;
                    pause1 = (int) (SystemClock.elapsedRealtime() - chrono.getBase());
                    lap_mili = (float) pause1 / 1000;
                    // will give an index by default
                    lapArrayList.add(index + ": college time:  " + lap_mili);
//                    lapList.add("study");
                    lapAdapter.notifyDataSetChanged();
                } else {
                    chrono.setBase(SystemClock.elapsedRealtime());
                    pauseOffset = 0;
                    chrono.stop();
                    // will clear of all the lap time recorded
                    lapAdapter.clear();
                    lapAdapter.notifyDataSetChanged();
                }

            }
        });

        clearLap = view.findViewById(R.id.clearLapFrag);
        clearLap.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                index = 0;
                lapAdapter.clear();
                lapAdapter.notifyDataSetChanged();
            }
        });

        return view;
    }

}

File: app/src/main/java/com/example/timemanagementapp/StopWatchFragment.kt
package com.example.timemanagementapp;

import android.annotation.SuppressLint;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Chronometer;
import android.widget.TextView;

import com.example.timemanagementapp.databinding.FragmentStopWatchBinding;

import java.time.chrono.ChronoLocalDateTime;

public class StopWatchFragment extends Fragment {

    public StopWatchFragment()
    {
        // constructor for fragment
    }

    private FragmentStopWatchBinding binding;
    private long timeInSeconds = 0;
    private boolean running = false;
    private Handler handler;
    private Runnable runnable;
    int count = 1;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // Example initialization of a button in the fragment's layout
        Button myButton = view.findViewById(R.id.resetButton);

    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_stop_watch, container, false);
        binding = FragmentStopWatchBinding.inflate(inflater, container, false);
        TextView timer = view.findViewById(R.id.timerTextView);
        Button start,stop,reset;
        start = view.findViewById(R.id.startButton);
        stop = view.findViewById(R.id.stopButton);
        reset = view.findViewById(R.id.resetButton);


        // handler and runnable is used to handle all the updates for the UI
        // it always keeps running in the background until it is stopped
        // chronometer also uses handler to update its UI
        // to handle stopwatch and countdown use handler and runnable it acts on the main thread and updates regularly without needed to be called
        handler = new Handler();
        runnable = new Runnable() {
            @Override
            public void run() {
                long hours = timeInSeconds / 3600;
                long minutes = (timeInSeconds % 3600) / 60;
                long seconds = timeInSeconds % 60;
                @SuppressLint("DefaultLocale") String timeString = String.format("%02d:%02d:%02d", hours, minutes, seconds);

                timer.setText(timeString);

                if (running) {
                    timeInSeconds++;
                }
                handler.postDelayed(this, 1000);
            }
        };


        start.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                running = true;
            }
        });

        stop.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                running = false;
            }
        });

        reset.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                running = false;
                timeInSeconds = 0;
                timer.setText("00:00:00");
            }
        });

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        handler.postDelayed(runnable, 1000);
    }

    @Override
    public void onPause() {
        super.onPause();
        handler.removeCallbacks(runnable);
    }

}
